### ISSUE TITLE
RFC: Unify conv/pool layers

### ISSUE THREAD

**Issue by craffel** [Lasagne/Lasagne#299]
It has come up a couple of times as I have introduced people (cc @bmcfee, @moscow25, and issue #128) to Lasagne that someone is trying to use some code which uses, for example, `Conv2DDNNLayer` on a machine which has no GPU or does not have cudnn installed (this can happen when you're running someone else's code, or when you want to test out some code you wrote on your GPU box on your GPU-less laptop).  The simple fix is generally to just use `Conv2DLayer` instead, but it's not really a portable solution to tell people to modify code when it's being used on different systems.  A simple hack fix would be for all scripts which use Lasagne's layers which have optional dependencies to have something like this (as an example) at the top:
```Python
try:
    import lasagne.layers.dnn
    conv_layer = lasagne.layers.dnn.Conv2DNNLayer
except ImportError:
    conv_layer = lasagne.layers.Conv2DLayer
...
# All convolutional layers would then be instantiated like this
layer = conv_layer(incoming, ...)
```
I don't really think it's fair to encourage all of our users to do something this hacky, though, and I think we owe it to users to abstract this out.  To me, a good solution would be for there to be a single `Conv2DLayer` which had an additional kwarg such as `backend`, which could be, say, `'theano'`, `'cuda_convnet'`, `'dnn'`, etc., which chose which backend to use.  Given that the different backends have different requirements, we'd need to be diligent (as we already are) about throwing errors when a backend can't support the provided layer arguments.  The simplest way to implement this would be to basically just make `Conv2DLayer` a wrapper class around all of these layers, which threw warnings when it fell back on the theano backend because of something not being present.  I believe this also applies to pooling layers.  If and when theano does this for us, we can greatly simplify our code, but I think it's worth abstracting this out now instead of waiting.  I don't have enough in-depth knowledge of the different layers/backends to know if there is something making this not possible, so please let me know if there is.

**Comment by ebenolson** [comment]
Another way to do this would be to add code to `cuda_convnet.py` and `dnn.py` which replaces those Layer classes with dummy wrappers around `Conv2DLayer` when the libraries are unavailable.

One complication is that `Conv2DLayer` doesn't support `pad` at the moment.

**Comment by benanne** [comment]
As I said in #298 already, the problem is that these layer implementations are not always interchangeable. They implement a different subset of features. Technically there should be no reason for `Conv2DDNNLayer` to exist: Theano can replace the default convolution used in `Conv2DLayer` with the cuDNN one where possible, so just using `Conv2DLayer` should be fine. Unfortunately the default Theano convolution interface is somewhat outdated and does not support things like arbitrary zero padding and same convolutions very well. That is why `Conv2DDNNLayer` exists. Falling back to `Conv2DLayer` simply won't always be possible.

Same with `Conv2DCCLayer`, except in this case the Theano optimizations that swap in the cuda-convnet ops don't exist (yet).

> To me, a good solution would be for there to be a single Conv2DLayer which had an additional kwarg such as backend, which could be, say, 'theano', 'cuda_convnet', 'dnn', etc.

This is the 'clean' solution, but take a look at those layer classes, some of them are monsters already. I really don't feel that we should make things even more complicated by merging them into one. The current situation with separate classes is a pragmatic solution to a problem that, as I said, should be solved at the Theano level.

I think merging them would be pretty confusing as well, as there would basically be two stages of 'backend resolution': a Lasagne stage, where we decide which Theano functions to call to implement the convolution (e.g. we could call `T.nnet.conv.conv2d`, or we could call `theano.sandbox.dnn.dnn_conv`) and then a Theano stage where it which ops to use to implement these functions (Theano might decide cuDNN to implement `T.nnet.conv.conv2d` anyway). That doesn't sound very elegant at all.

> I think it's worth abstracting this out now instead of waiting. 

I don't think it is, to be honest. There are plenty of other things that would be more useful to spend our time on. We would essentially be doing double work as there are plans to fix this at the Theano level, which is where it should be fixed. Admittedly we may be waiting a while for that... but maybe then we should be contributing there instead.

**Comment by benanne** [comment]
> Another way to do this would be to add code to `cuda_convnet.py` and `dnn.py` which replaces those Layer classes with dummy wrappers around `Conv2DLayer` when the libraries are unavailable.

I don't think that's a good idea. If you're using `Conv2DCCLayer` in your code, i.e. explicitly instructing Lasagne to use the `cuda_convnet` implementation, it should not quietly go and use a different one if that one isn't available, it should fail.

**Comment by craffel** [comment]
> Another way to do this would be to add code to cuda_convnet.py and dnn.py which replaces those Layer classes with dummy wrappers around Conv2DLayer when the libraries are unavailable.

Yeah, that would work.  I can think of a few ways to do this with minimal overhead which may or may not be best - e.g., defining a function `Conv2DLayer` which returns an instance of the correct underlying class given the value of a new `backend` keyword argument, with correct fallbacks.  It would look like a class constructor to the end-user.  Here's an incomplete example of what I mean:

```Python
from theano.sandbox.cuda import dnn
if not theano.config.device.startswith("gpu") or not dnn.dnn_available():
    HAS_DNN = True
else:
    HAS_DNN = False

def Conv2DLayer(incoming, num_filters, filter_size, backend='theano', **kwargs):
    if backend == 'cudnn':
        if HAS_DNN:
            return Conv2DDNNLayer(incoming, num_filters, filter_size, **kwargs)
        else:
            warnings.warn("backend='dnn' was provided, but cudnn is not available.")
            backend = 'theano'
    if backend == 'theano':
            # We'd need to rename the current Conv2DLayer, this is a bad name
            return Conv2DTheanoLayer(incoming, num_filters, filter_size, **kwargs)
```
etc.

> One complication is that Conv2DLayer doesn't support pad at the moment.

and

> the problem is that these layer implementations are not always interchangeable. They implement a different subset of features. 

Right, but I think it's definitely preferable to have them just throw errors when there is a feature that is requested but not available (padding, or non-square filters which I believe aren't supported by `cuda_convnet`, etc.)  Ideally, we'd have a backend which works for all kwargs all the time, which I believe the vanilla theano backend is aside from padding, right?  But even without that, I'd prefer this.

> This is the 'clean' solution, but take a look at those layer classes, some of them are monsters already. I really don't feel that we should make things even more complicated by merging them into one.

Let me know what you think of the above alternative - less clean, but with very little code added to the library.

> That doesn't sound very elegant at all.

It's not, but theano is forcing us to be this way.

> I don't think it is, to be honest. There are plenty of other things that would be more useful to spend our time on.

I don't think it would take much time to implement something simple, although I'm not convinced the simple solution is best.

> We would essentially be doing double work as there are plans to fix this at the Theano level, which is where it should be fixed.

If we do wait for Theano, we need to be very careful about documenting this - i.e., rather than `raise ImportError("dnn not available")`, have the error message be something like "lasagne.layers.dnn was imported, but cudnn is not available.  The code you are running will only work on a computer with a GPU and with cudnn installed.  If you'd like to use this code on another system, you need to change all instances of Conv2DDNNLayer to Conv2DLayer and all instances of MaxPool2DDNNLayer to MaxPool2DLayer (etc etc)"

> it should not quietly go and use a different one if that one isn't available, it should fail.

It absolutely should not go quietly!  There should be a warning.  I don't think it should fail though.  A good example of this is the way `librosa` handles its dependency on `scikits.samplerate`; when `resample` is called and `scikits.samplerate` is not available, it will fall back on `scipy`'s resample (which is slower), but the user is warned about it.

To summarize, the reason this is frustrating to me is that as it is, Lasagne (by way of Theano) basically forces you to write non-portable code when you want to use one of the "better" backends.  But, with some relatively small changes to the codebase, we could make Lasagne code transparently (with error messages) portable up to the inherent deficiencies of each individual backend (with clear error messages in these cases).

**Comment by ebenolson** [comment]
> I don't think that's a good idea. If you're using Conv2DCCLayer in your code, i.e. explicitly instructing Lasagne to use the cuda_convnet implementation, it should not quietly go and use a different one if that one isn't available, it should fail.

Well, it would definitely need to raise a warning of some sort, but I think it would be preferable to adding all the substitution complexity to `Conv2DLayer`. No changes to the API or existing user code would be necessary, it would just provide a way to let code run (suboptimally) even if the desired libraries aren't available.

But yeah I'm not convinced it's necessary, for me it works pretty well to do e.g.

    from lasagne.layers.dnn import Conv2DDNNLayer as ConvLayer

which makes swapping the backend pretty painless in most cases.

**Comment by craffel** [comment]
> But yeah I'm not convinced it's necessary, for me it works pretty well to do e.g.
from lasagne.layers.dnn import Conv2DDNNLayer as ConvLayer
which makes swapping the backend pretty painless in most cases.

This still isn't portable.  You'd need to do (in every python file you write which uses the cudnn backend...)
```Python
try:
    import lasagne.layers.dnn
    ConvLayer = lasagne.layers.dnn.Conv2DNNLayer
except ImportError:
    ConvLayer = lasagne.layers.Conv2DLayer
```

**Comment by nouiz** [comment]
Hope fully or will be fixed next week in theano! But don't be surprised if
there is a little bit of extra delay...

We where close to finish it but nips delayed it.

**Comment by benanne** [comment]
Disclaimer: everything that follows is my personal opinion and not "the way it will be" :) As it turns out I feel quite strongly about this issue so I apologize if anything I say comes across as harsh.

> If we do wait for Theano, we need to be very careful about documenting this - i.e., rather than raise ImportError("dnn not available"), have the error message be something like "lasagne.layers.dnn was imported, but cudnn is not available. The code you are running will only work on a computer with a GPU and with cudnn installed. If you'd like to use this code on another system, you need to change all instances of Conv2DDNNLayer to Conv2DLayer and all instances of MaxPool2DDNNLayer to MaxPool2DLayer (etc etc)"

I think that would be preferable. Any solution that somehow merges these different implementations seems overkill and unnecessarily complicated to me. But if someone can come up with something reasonably elegant I may be persuaded. I don't know.

> It absolutely should not go quietly! There should be a warning. I don't think it should fail though. A good example of this is the way librosa handles its dependency on scikits.samplerate; when resample is called and scikits.samplerate is not available, it will fall back on scipy's resample (which is slower), but the user is warned about it.

A warning is easy to miss. The backends do not only support different subsets of features, their behavior is also different (some configurations might be way, way slower with certain fallback implementations) and the numerical rounding behavior is different as well, so it would potentially lead to different results. If your script starts printing epoch info straight after, you will be guaranteed to miss that warning as well.

We should simply raise an error and explain to the user that they can change the backend if they wish. We should not do this for them automatically. I think that would be a mistake. 

I don't think the librosa example is relevant: you don't explicitly tell librosa "now resample this using scikits.samplerate", you tell it "resample this please" and you don't really care how it happens. If you use `Conv2DCCLayer` or `Conv2DLayer(backend='cuda_convnet')` or whatever, you are saying, "use this particular implementation". We should not try to be too "helpful" if that particular implementation is unavailable. Most likely they would prefer to do what's necessary to make it available over using a different backend anyway.

> To summarize, the reason this is frustrating to me is that as it is, Lasagne (by way of Theano) basically forces you to write non-portable code when you want to use one of the "better" backends.

It's unfortunate, but as I said, it is an issue that should be solved in Theano. Any solution on the Lasagne level is inelegant and adds unnecessary complexity, something we are trying really hard to avoid. It's better to be transparent about these shortcomings than to try to do "smart" things to cover them up.

> it would just provide a way to let code run (suboptimally) even if the desired libraries aren't available.

There is already a way: if the code doesn't run, swap out one implementation for another. It will only take a minute or two. As before, I really don't think this is something that needs to be automated. It needs to be a conscious decision on the developer's part.

> Hope fully or will be fixed next week in theano! But don't be surprised if
there is a little bit of extra delay...

Ahh, there we go :) That sounds like the preferred way to go - provided that this unified interface will be available when using the 'old' backend as well, and not just the new one.

**Comment by nouiz** [comment]
On Fri, Jun 12, 2015 at 6:54 PM, Sander Dieleman <notifications@github.com>
wrote:

[...]

Hope fully or will be fixed next week in theano! But don't be surprised if
>
> there is a little bit of extra delay...
>
> Ahh, there we go :) That sounds like the preferred way to go - provided
> that this unified interface will be available when using the 'old' backend
> as well, and not just the new one.
>
I'm not sure I understand correctly, but I think that
theano.tensor.nnet.conv.conv2d() will just be replaced with the new
operation and should support the pad fonctionality as cudnn and gpucorrmm.
So people could use your conv2dlayer interface that do not specify which
implementation. But that won't trigger cuda_convnet implementation.

Adding optimization in pylearn2 to use that implementation isn't too hard,
but as they have different memory layout, this can loose speed. Did you
time cuda convnet vs gpucorrmm on GTX580? I think it is the only GPU where
cuda convnet could be usefull. On kepler and fermi GPU, I'm pretty sure
cudnn is faster then cuda convnet.

**Comment by benanne** [comment]
> I'm not sure I understand correctly, but I think that
theano.tensor.nnet.conv.conv2d() will just be replaced with the new
operation and should support the pad fonctionality as cudnn and gpucorrmm.
So people could use your conv2dlayer interface that do not specify which
implementation.

Great! That would be the plan indeed. We could at least do away with our direct corrmm and cudnn wrappers this way.

> But that won't trigger cuda_convnet implementation.

True, we will still need to wrap that separately. I would argue in favour of keeping this as a separate class, because of the different memory layout (which we handle with an optional `dimshuffle` flag) and the various limitations associated with it.

> Did you time cuda convnet vs gpucorrmm on GTX580? I think it is the only GPU where
cuda convnet could be usefull. On kepler and fermi GPU, I'm pretty sure
cudnn is faster then cuda convnet.

I don't have a GTX 580 on hand. But I know for a fact that cuda-convnet (v1, the one wrapped in pylearn2) is faster than CuDNN v2 on the GTX 980 for some configurations (like 20-40% faster, so it's definitely significant). I think this is because neither are optimized for Maxwell, and the cuda-convnet code just happens to translate better. (It's possible that cuda-convnet v2 is even faster, but that isn't available in Theano right now.)

This will most likely change when CuDNN v3 is released, since Maxwell optimizations are something they are supposedly working on.

**Comment by craffel** [comment]
> As it turns out I feel quite strongly about this issue so I apologize if anything I say comes across as harsh.

No worries, it's good to hash this out!  I think the reason I feel strongly about it is that I got some pretty strong reactions when trying to explain to people why they had to change `Conv2DDNNLayer` in my code to `Conv2DLayer` to get it running.

Granted, much of this discussion is sort of moot if Theano makes the different backends transparently swappable.

> But if someone can come up with something reasonably elegant I may be persuaded. I don't know.

I think what I posted is reasonably elegant, or at least, reasonably simple and gets the job done with no overhead.

> A warning is easy to miss.

Arguably if we show the user a warning it's up to them to notice.  It's better, IMO, then throwing an error and having them give up, which I'm sure has happened.

> The backends do not only support different subsets of features,

Right, but that's why we throw errors when a feature isn't available.

> their behavior is also different (some configurations might be way, way slower with certain fallback implementations) and the numerical rounding behavior is different as well, so it would potentially lead to different results.

Seriously?  This seems kind of crazy - shouldn't they all just compute convolutions?  And if not, isn't the idea that we created "functionally equivalent layers up to the extent possible"?  To whatever extent this is true, I think it needs to be heavily documented, as even as the code is I think people will assume the different layers will do the same thing.

> I don't think the librosa example is relevant: you don't explicitly tell librosa "now resample this using scikits.samplerate", you tell it "resample this please" and you don't really care how it happens.

I guess what I'm not seeing here is how resampling (which will also have slightly different numerical results, but is fundamentally performing the same operation) is any different from convolution (a pretty precisely defined mathematical operation).  When you construct a convolutional layer, you're saying "convolve this please (with this # filters, kernel size, padding scheme, etc)" and you don't really care how it happens.  Maybe I am the one who needs the aforementioned documentation of how the different backends are different the most.

> Most likely they would prefer to do what's necessary to make it available over using a different backend anyway.

Not in my experience.  If someone doesn't have a GPU, they aren't going to buy a GPU box.

> it is an issue that should be solved in Theano.

Absolutely agree, and if that happens soon that's great.

> It's better to be transparent about these shortcomings than to try to do "smart" things to cover them up.

I would argue that we are not being transparent about the shortcomings, and that we already are doing "smart" things to make the layers as functionally equivalent as possible.

**Comment by nouiz** [comment]
Le 13 juin 2015 14:55, "Colin Raffel" <notifications@github.com> a écrit :
>>
>> As it turns out I feel quite strongly about this issue so I apologize if
anything I say comes across as harsh.
>
> No worries, it's good to hash this out! I think the reason I feel
strongly about it is that I got some pretty strong reactions when trying to
explain to people why they had to change Conv2DDNNLayer in my code to
Conv2DLayer to get it running.
>
> Granted, much of this discussion is sort of moot if Theano makes the
different backends transparently swappable.
>>
>> But if someone can come up with something reasonably elegant I may be
persuaded. I don't know.
>
> I think what I posted is reasonably elegant, or at least, reasonably
simple and gets the job done with no overhead.
>>
>> A warning is easy to miss.
>
> Arguably if we show the user a warning it's up to them to notice. It's
better, IMO, then throwing an error and having them give up, which I'm sure
has happened.
>>
>> The backends do not only support different subsets of features,
>
> Right, but that's why we throw errors when a feature isn't available.
>>
>> their behavior is also different (some configurations might be way, way
slower with certain fallback implementations) and the numerical rounding
behavior is different as well, so it would potentially lead to different
results.
>
> Seriously? This seems kind of crazy - shouldn't they all just compute
convolutions? And if not, isn't the idea that we created "functionally
equivalent layers up to the extent possible"? To whatever extent this is
true, I think it needs to be heavily documented, as even as the code is I
think people will assume the different layers will do the same thing.
>>
>> I don't think the librosa example is relevant: you don't explicitly tell
librosa "now resample this using scikits.samplerate", you tell it "resample
this please" and you don't really care how it happens.
>
> I guess what I'm not seeing here is how resampling (which will also have
slightly different numerical results, but is fundamentally performing the
same operation) is any different from convolution (a pretty precisely
defined mathematical operation). When you construct a convolutional layer,
you're saying "convolve this please (with this # filters, kernel size,
padding scheme, etc)" and you don't really care how it happens. Maybe I am
the one who needs the aforementioned documentation of how the different
backends are different the most.
>>
>> Most likely they would prefer to do what's necessary to make it
available over using a different backend anyway.
>
> Not in my experience. If someone doesn't have a GPU, they aren't going to
buy a GPU box.

I get frequently remark in theano mailing listed that people bought a GPU
to get it much faster. The GPU do not only enable the computation, it make
it much faster. It happen frequently that people all recommendation on the
fly configuration to buy.

**Comment by craffel** [comment]
> I get frequently remark in theano mailing listed that people bought a GPU
to get it much faster. The GPU do not only enable the computation, it make
it much faster. It happen frequently that people all recommendation on the
fly configuration to buy.

Right, maybe I wasn't being clear - many people (myself included) have purchased GPUs to make their networks run faster, but I doubt anyone gets the Lasagne error "dnn not available" and rushes to the store to buy a GPU box so that they can get the code to run.

**Comment by benanne** [comment]
> I think what I posted is reasonably elegant, or at least, reasonably simple and gets the job done with no overhead.

If we can still split the code across different modules somehow, it might be doable to have a `backend` keyword argument. Incidentally there already is a `convolution` keyword argument which sort of serves the same purpose, but it expects a function that mimics the interface of `T.nnet.conv.conv2d`, so it is useless in combination with the other implementations supported by Theano / pylearn2 (they all have different interfaces).

> Arguably if we show the user a warning it's up to them to notice.  It's better, IMO, then throwing an error and having them give up, which I'm sure has happened.

I think this will be at least as annoying as the current situation ;)

> > their behavior is also different (some configurations might be way, way slower with certain fallback implementations) and the numerical rounding behavior is different as well, so it would potentially lead to different results.

> Seriously?  This seems kind of crazy - shouldn't they all just compute convolutions?  And if not, isn't the idea that we created "functionally equivalent layers up to the extent possible"?  To whatever extent this is true, I think it needs to be heavily documented, as even as the code is I think people will assume the different layers will do the same thing.

The thing is that training a neural net is an iterative process, so small numerical differences start to matter a lot after a few 100 iterations. You can get completely different end results using different convolution implementations with the exact same initialization. In fact, due to the nondeterminism of the gradient in cuDNN, you can get completely different results with two subsequent runs (with the exact same initialization). So for reproducibility's sake, swapping implementations when not explicitly instructed to do so by the user is a bad idea.

> I guess what I'm not seeing here is how resampling (which will also have slightly different numerical results, but is fundamentally performing the same operation) is any different from convolution (a pretty precisely defined mathematical operation).  When you construct a convolutional layer, you're saying "convolve this please (with this # filters, kernel size, padding scheme, etc)" and you don't really care how it happens.  Maybe I am the one who needs the aforementioned documentation of how the different backends are different the most.

But the thing is, if you are using `Conv2DDNNLayer`, `Conv2DMMLayer` or `Conv2DCCLayer`, you *do* care how it happens - that's why you are using them in the first place. If you didn't care, you could just use `Conv2DLayer` and let it 'optimize' in the convolution backend of your choice. The alternative implementations exist precisely to give the user the option to go 'low-level' and manually select how things are implemented.

Currently, one exception to this is of course the fact that people resort to `Conv2DDNNLayer` because they want to use a feature that the default layer does not support (i.e. arbitrary padding). We should probably just implement this on `Conv2DLayer` as well so that doesn't occur anymore. Hopefully the new Theano convolution interface will already take care of this.

> > Most likely they would prefer to do what's necessary to make it available over using a different backend anyway.

> Not in my experience.  If someone doesn't have a GPU, they aren't going to buy a GPU box.

I wasn't thinking about buying hardware, but about installing missing dependencies (e.g. CuDNN, pylearn2). Honestly, if someone doesn't have a GPU and they want to train convnets, they should think twice about what they are doing :P But that is besides the point of course.

What I want to avoid is that people end up using a slow implementation while they think CuDNN is being used, because CuDNN was not installed properly. Things like that. I don't think warnings suffice. A good example of this problem is Theano itself: if you use `T.nnet.conv.conv2d`, it should use CuDNN if it is installed. If not, it will fall back to GpuCorrMM. It's happened to me more than once that I was 'accidentally' using GpuCorrMM because some sort of path wasn't set correctly. Quite annoying! (In fairness, Theano doesn't even generate a warning. But if there was a warning, I'm sure I would miss it the first few times I'm looking.)

> I would argue that we are not being transparent about the shortcomings, and that we already are doing "smart" things to make the layers as functionally equivalent as possible.

I think it's as transparent and 'dumb' as can be: `Conv2DLayer` wraps `T.nnet.conv.conv2d`, the others wrap specialized implementations. The problem is probably that there is too much of an incentive to use the specialized implementations right now. They were really intended to be used only by 'advanced users', so to speak. The new Theano convolution interface should change that at least.


---

Anyway, we are sort of mixing up two separate proposals:
1. unify the convolution interfaces so all implementations can be accessed using a single class
2. make this unified interface automatically 'fall back' if a chosen backend is not available

If #1 is done properly (i.e. not just throwing all of the code in a single class, which would be a completely unreadable monster), I could get behind that. I think #2 is a bad idea though.

One compromise for #2 could be that you could manually enable this 'fallback' behavior with another flag. It should be off by default though.

**Comment by craffel** [comment]
> You can get completely different end results using different convolution implementations with the exact same initialization.

I guess I didn't realize it was this bad.  This seems problematic to me in general, but not something we can necessarily fix.

> But the thing is, if you are using Conv2DDNNLayer, Conv2DMMLayer or Conv2DCCLayer, you do care how it happens - that's why you are using them in the first place. If you didn't care, you could just use Conv2DLayer and let Theano 'optimize' in the convolution backend of your choice. 

Wait - sorry, there may be a big misunderstanding here - if I have cudnn installed, and I use a `Conv2DLayer`, will Theano use cudnn?  Because my understanding was that `Conv2DLayer` used theano's native slow convolution, and that if I wanted to use cudnn, I had to use `Conv2DDNNLayer`.  The reason I used `Conv2DDNNLayer`, for example, was simply because I thought it was the only way I could use cudnn and because it would be the fastest, not because I care about the implementation at all.  If I've misunderstood here, sorry! - but I am also certain I'm not the only person who thinks it works this way...!

> A good example of this problem is Theano itself: if you use T.nnet.conv.conv2d, it should use CuDNN if it is installed. If not, it will fall back to GpuCorrMM. It's happened to me more than once that I was 'accidentally' using GpuCorrMM because some sort of path wasn't set correctly. Quite annoying! (In fairness, Theano doesn't even generate a warning. But if there was a warning, I'm sure I would miss it the first few times I'm looking.)

This seems worse than annoying if indeed the results can be substantially different.

> If #1 is done properly (i.e. not just throwing all of the code in a single class, which would be a completely unreadable monster), I could get behind that. I think #2 is a bad idea though.

Yeah, the first option is enough for me as long as there is a very simple thing the user can do (preferably only modifying code in a single place) to manually set the fall-back.

**Comment by benanne** [comment]
> Wait - sorry, there may be a big misunderstanding here - if I have cudnn installed, and I use a `Conv2DLayer`, will Theano use cudnn?  Because my understanding was that `Conv2DLayer` used theano's native slow convolution, and that if I wanted to use cudnn, I had to use `Conv2DDNNLayer`.  The reason I used `Conv2DDNNLayer`, for example, was simply because I thought it was the only way I could use cudnn and because it would be the fastest, not because I care about the implementation at all. 

Now it all starts to make sense :) It's as you say, if you use `Conv2DLayer` and CuDNN is installed, CuDNN will be used. Unfortunately Theano's `T.nnet.conv.conv2d` does not support arbitrary zero padding / same border mode, so if you want those things, it's better to use `Conv2DDNNLayer`, which calls `dnn_conv` directly and does support those things properly. That's the only good reason to use it. Otherwise you should go with `Conv2DLayer`. It will use the same backend (provided everything is configured correctly). Hopefully all of this will be moot with the new convolution interface though.

> If I've misunderstood here, sorry! - but I am also certain I'm not the only person who thinks it works this way...!

Yep, sounds like this is worth clarifying in the docs as well. The misunderstanding seems to stem from the fact that Theano has both the user-facing `T.nnet.conv.conv2d` interface, for which there are several implementations, as well as specific interfaces for some of those implementations (which are sometimes very different).

**Comment by craffel** [comment]
> Now it all starts to make sense :) It's as you say, if you use Conv2DLayer and CuDNN is installed, CuDNN will be used. Unfortunately Theano's T.nnet.conv.conv2d does not support arbitrary zero padding / same border mode, so if you want those things, it's better to use Conv2DDNNLayer, which calls dnn_conv directly and does support those things properly. That's the only good reason to use it. Otherwise you should go with Conv2DLayer. It will use the same backend (provided everything is configured correctly). Hopefully all of this will be moot with the new convolution interface though.

Damn, ok, yeah, I was completely misunderstanding, and I think most of the non-dev people I knew who are using Lasagne understand it the way I saw it too.  

> Yep, sounds like this is worth clarifying in the docs as well. The misunderstanding seems to stem from the fact that Theano has both the user-facing T.nnet.conv.conv2d interface, for which there are several implementations, as well as specific interfaces for some of those implementations (which are sometimes very different).

Yeah, we should clarify this.  I think for me it was an incorrect assumption based on not understanding Theano's fallback behavior - i.e. I didn't know Theano would ever use cudnn, so it made sense to me that you created a separate cudnn wrapper layer for using it.  But, that was my mistake.  I'll close this now, perhaps we can create a separate issue to clarify this and/or update the layers when Theano's interface is unified.

**Comment by benanne** [comment]
Sounds good!

**Comment by f0k** [comment]
> Hope fully or will be fixed next week in theano!

@nouiz: Is that https://github.com/Theano/Theano/pull/2665 or will there be a new PR? If it's okay with you, I'd like to review the interface change as well, please keep me updated!

> (It's possible that cuda-convnet v2 is even faster, but that isn't available in Theano right now.)

That's something we could change ourselves!

> Yeah, we should clarify this. I think for me it was an incorrect assumption based on not understanding Theano's fallback behavior - i.e. I didn't know Theano would ever use cudnn, so it made sense to me that you created a separate cudnn wrapper layer for using it.

The Theano documentation explains quite well that `conv2d()` will be replaced by other implementations: http://deeplearning.net/software/theano/library/tensor/nnet/conv.html
But if you don't know Theano before, the mere existence of the extra classes in Lasagne suggest that you'd need them to change the implementation. So I'm in favor of ditching the extra classes once `theano.tensor.conv2d()` is adapted (except for the cuda-convnet one -- not because it doesn't have an optimizer, we could add that to pylearn2, but because of the different memory layout).

**Comment by craffel** [comment]
> The Theano documentation explains quite well that conv2d() will be replaced by other implementations: http://deeplearning.net/software/theano/library/tensor/nnet/conv.html
But if you don't know Theano before, the mere existence of the extra classes in Lasagne suggest that you'd need them to change the implementation.

Yes, that was exactly my error.

**Comment by nouiz** [comment]
Le 16 juin 2015 13:55, "Jan Schlüter" <notifications@github.com> a écrit :
>>
>> Hope fully or will be fixed next week in theano!
>
> @nouiz: Is that Theano/Theano#2665 or will there be a new PR? If it's
okay with you, I'd like to review the interface change as well, please keep
me updated!

I'm offline now, but it is probably this one. It is about meta conv or
abstract conv. There is a few corner case not working yet. You are welcome
to review it! Don't hesitate (or anyone else)

>>
>> (It's possible that cuda-convnet v2 is even faster, but that isn't
available in Theano right now.)
>
> That's something we could change ourselves!

It should be faster on titan from memory, but it have a separate license
that make we must put it in a different repo. I don't know anyone planning
to work on this, so if you are interested, don't hesitate to  work on it.
Cuda convnet v1 wrapper should be a good start.

>>
>> Yeah, we should clarify this. I think for me it was an incorrect
assumption based on not understanding Theano's fallback behavior - i.e. I
didn't know Theano would ever use cudnn, so it made sense to me that you
created a separate cudnn wrapper layer for using it.
>
> The Theano documentation explains quite well that conv2d() will be
replaced by other implementations:
http://deeplearning.net/software/theano/library/tensor/nnet/conv.html
> But if you don't know Theano before, the mere existence of the extra
classes in Lasagne suggest that you'd need them to change the
implementation. So I'm in favor of ditching the extra classes once
theano.tensor.conv2d() is adapted (except for the cuda-convnet one -- not
because it doesn't have an optimizer, we could add that to pylearn2, but
because of the different memory layout).

**Comment by benanne** [comment]
> That's something we could change ourselves!

I'd love to do some hacking again, but I'm stuck writing prose instead :p

> So I'm in favor of ditching the extra classes once `theano.tensor.conv2d()` is adapted (except for the cuda-convnet one -- not because it doesn't have an optimizer, we could add that to pylearn2, but because of the different memory layout).

Sounds good, as long as we don't lose any functionality or performance improvements (iirc there can be a difference between using the optimizer for CuDNN and using `dnn_conv` directly).

**Comment by craffel** [comment]
(2 years later...)

>  Unfortunately Theano's T.nnet.conv.conv2d does not support arbitrary zero padding / same border mode, so if you want those things, it's better to use Conv2DDNNLayer, which calls dnn_conv directly and does support those things properly.

What is the status of this?  Have these differences been resolved?  I.e. is there any reason to use Conv2DDNNLayer instead of just Conv2DLayer any more?  Asking because I was looking at this code (from a prominent research project, no less!) and they are using Conv2DDNNLayer:
https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/train_cifar_feature_matching.py#L56

**Comment by nouiz** [comment]
Using directly the DNN layers isn't needed anymore with a recent Theano.

But changing people habit take times. We will continue to see this for a
long time probably.

**Comment by f0k** [comment]
@craffel: Nice to see you again :)

> But changing people habit take times.

We could also deprecate the layers, but maybe it's handy to be able to use a specific implementation for specific layers sometimes.

**Comment by craffel** [comment]
> We could also deprecate the layers, but maybe it's handy to be able to use a specific implementation for specific layers sometimes.

Can you think of any reason at this point someone would want to force use of cudnn kernels, now that they offer no additional concrete functionality?  They seem to create a lot of confusion.

A simple way to deprecate them without breaking code would be to make them aliases for non-DNN layers (e.g. just setting `Conv2DDNNLayer = Conv2DLayer` etc. in the dnn submodule).  Or if you want to be fancy, `Conv2DDNNLayer` could subclass `Conv2DLayer` and just raise a warning before doing `super(...).__init__` and that's it.

**Comment by f0k** [comment]
> Can you think of any reason at this point someone would want to force use of cudnn kernels, now that they offer no additional concrete functionality?

I was thinking of explicit mix-and-matching, like using FFT-based convolution for the first few layers, and gemm-based for the rest, or deterministic convolution for some layers, and "anything goes" for the rest. But probably nobody will ever need this (or not as part of the core library).

Let's phase out the implementation-specific modules after the 0.2 release, since at that point we can also drop the BatchNormDNNLayer (that'd be complicated as long as we support Theano 0.8).

> A simple way to deprecate them without breaking code would be to make them aliases for non-DNN layers

I think there were some differences in the default values somewhere. So we'll have to go the fancier route.
